### PR TITLE
check that tree math input is valid

### PR DESCRIPTION
**Pull request by franziskuskiefer**
I ran into some infinite loops while testing because tree math didn't check that inputs are valid. I added a couple checks that should cover most cases.

### CODE REVIEW

File: src/tree/treemath.rs [hunk 1]
```diff
@@ -69,40 +72,73 @@ pub(crate) fn right(index: NodeIndex, size: LeafIndex) -> Result<NodeIndex, Tree
...
+fn unsafe_parent(index: NodeIndex, size: LeafIndex) -> Result<NodeIndex, TreeMathError> {
     let x = index.as_usize();
     let n = size.as_usize();
     if index == root(size) {
         return Err(TreeMathError::RootHasNoParent);
     }
     let mut p = parent_step(x);
-    while p >= node_width(n) {
-        p = parent_step(p)
+    let node_width = node_width(n);
```
Comment (kkohbrok) [inline src/tree/treemath.rs:96]: I'm just curious: Why the explicit `node_width` variable here?

File: src/tree/treemath.rs [hunk 2]
```diff
@@ -69,40 +72,73 @@ pub(crate) fn right(index: NodeIndex, size: LeafIndex) -> Result<NodeIndex, Tree
...
+    while p >= node_width {
+        let new_p = parent_step(p);
+        if new_p == p {
```
Comment (kkohbrok) [inline src/tree/treemath.rs:99]: Can `parent_step` of an index be the same as the index itself? I thought we should be checking here if the output is outside of the tree.

File: src/tree/treemath.rs [hunk 3]
```diff
@@ -69,40 +72,73 @@ pub(crate) fn right(index: NodeIndex, size: LeafIndex) -> Result<NodeIndex, Tree
...
 }
 
 pub(crate) fn parent(index: NodeIndex, size: LeafIndex) -> Result<NodeIndex, TreeMathError> {
+    node_in_tree(index, size)?;
+    unsafe_parent(index, size)
+}
+
+// This function is only safe to use if index <= size.
+// If this is not checked before calling the function, `parent` should be used.
+fn unsafe_parent(index: NodeIndex, size: LeafIndex) -> Result<NodeIndex, TreeMathError> {
```
Comment (kkohbrok) [inline src/tree/treemath.rs:89]: If this function returns `Result` anyway, why not just have it return an error if the input index is outside of the tree?

**Review by kkohbrok — COMMENTED** [review]
Thanks for picking this up! Just a few questions, otherwise it should be good to go.

File: src/tree/treemath.rs (hunk 1 above)
Comment (franziskuskiefer) [reply src/tree/treemath.rs:96]: I guess it's a debug artefact that I forgot to remove...

File: src/tree/treemath.rs (hunk 3 above)
Comment (franziskuskiefer) [reply src/tree/treemath.rs:89]: This is a pure "performance improvement". It's avoiding the check when we know that it was done before. (I don't think we'd notice the performance difference...)

File: src/tree/treemath.rs (hunk 2 above)
Comment (franziskuskiefer) [reply src/tree/treemath.rs:99]: Well, this is a purely opportunistic check. If this ends up being true, we'd end up in an endless loop. It's not an actual correctness check.

**Review by kkohbrok — DISMISSED** [review]
Let's get this in!

File: src/tree/treemath.rs [hunk 4]
```diff
@@ -69,40 +72,72 @@ pub(crate) fn right(index: NodeIndex, size: LeafIndex) -> Result<NodeIndex, Tree
...
+    let p = unsafe_parent(index, size)?;
     match index.cmp(&p) {
         Ordering::Less => right(p, size),
         Ordering::Greater => left(p),
         Ordering::Equal => left(p),
     }
 }
 
+#[inline(always)]
+fn node_in_tree(node_index: NodeIndex, size: LeafIndex) -> Result<(), TreeMathError> {
```
Comment (raphaelrobert) [inline src/tree/treemath.rs:117]: This function should have a dedicated test.

File: src/tree/treemath.rs [hunk 5]
```diff
@@ -69,40 +72,72 @@ pub(crate) fn right(index: NodeIndex, size: LeafIndex) -> Result<NodeIndex, Tree
...
+    if node_index.as_usize() >= node_width(size.as_usize()) {
+        Err(TreeMathError::NodeNotInTree)
+    } else {
+        Ok(())
+    }
+}
+
+#[inline(always)]
+fn leaf_in_tree(leaf_index: LeafIndex, size: LeafIndex) -> Result<(), TreeMathError> {
```
Comment (raphaelrobert) [inline src/tree/treemath.rs:126]: This function should have a dedicated test.

**Review by raphaelrobert — CHANGES_REQUESTED** [review]
Thanks! LGTM in general, I just flagged two missing tests that I think should be there.

**Review by raphaelrobert — COMMENTED** [review]
Thanks for adding the tests! I think codecov is unhappy because you used `.unwrap()` instead of `.expect()`.

**Review by raphaelrobert — APPROVED** [review]

**Pull request merged**